### PR TITLE
Feature/custom auth

### PR DESCRIPTION
**Pull request by chillaq**

### CODE REVIEW

File: client/src/main/java/io/split/client/HttpSplitChangeFetcher.java [hunk 1]
```diff
@@ -38,16 +35,16 @@ public final class HttpSplitChangeFetcher implements SplitChangeFetcher {
     private static final String HEADER_CACHE_CONTROL_NAME = "Cache-Control";
     private static final String HEADER_CACHE_CONTROL_VALUE = "no-cache";
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/HttpSplitChangeFetcher.java:36]: please, remove both strings, not using them anymore.

File: client/src/main/java/io/split/client/RequestDecorator.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package io.split.client;
+
+import org.apache.hc.client5.http.classic.methods.HttpUriRequestBase;
+import java.util.*;
+
+class NoOpHeaderDecorator implements  UserCustomHeaderDecorator {
+    public NoOpHeaderDecorator() {}
+    @Override
+    public Map<String, String> getHeaderOverrides() {
+        return new HashMap<String, String>();
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/RequestDecorator.java:10]:
```suggestion
        return new HashMap<>();
```

File: client/src/main/java/io/split/client/SplitClientConfig.java [hunk 3]
```diff
@@ -10,11 +10,8 @@
 import pluggable.CustomStorageWrapper;
 
 import java.io.IOException;
-import java.util.HashSet;
-import java.util.LinkedHashSet;
-import java.util.List;
+import java.util.*;
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/SplitClientConfig.java:13]:
```suggestion
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.ThreadFactory;
```

Comment (chillaq) [reply client/src/main/java/io/split/client/SplitClientConfig.java:13]: I found these imports are required:
import java.util.HashSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;
import java.util.Set;

File: client/src/main/java/io/split/client/HttpSplitChangeFetcher.java (hunk 1 above)
Comment (chillaq) [reply client/src/main/java/io/split/client/HttpSplitChangeFetcher.java:36]: will do

File: client/src/main/java/io/split/client/UserCustomHeaderDecorator.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package io.split.client;
+
+import java.util.*;
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/UserCustomHeaderDecorator.java:3]:
```suggestion
import java.util.Map;
```

Comment (chillaq) [reply client/src/main/java/io/split/client/UserCustomHeaderDecorator.java:3]: will do

File: client/src/main/java/io/split/client/events/EventsSender.java [hunk 5]
```diff
@@ -18,20 +18,20 @@ public class EventsSender {
 
     private static final String BULK_ENDPOINT_PATH = "api/events/bulk";
     private final URI _bulkEndpoint;
-    private final CloseableHttpClient _client;
+    private final SplitHttpClient _client;
     private final TelemetryRuntimeProducer _telemetryRuntimeProducer;
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/events/EventsSender.java:22]:
```suggestion
```

File: client/src/main/java/io/split/client/impressions/HttpImpressionsSender.java [hunk 6]
```diff
@@ -63,32 +63,26 @@ private HttpImpressionsSender(CloseableHttpClient client, URI impressionBulkTarg
     @Override
     public void postImpressionsBulk(List<TestImpressions> impressions) {
 
-        CloseableHttpResponse response = null;
+        SplitHttpResponse response = null;
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/client/impressions/HttpImpressionsSender.java:66]:
```suggestion
        SplitHttpResponse response;
```

Comment (chillaq) [reply client/src/main/java/io/split/client/impressions/HttpImpressionsSender.java:66]: will do

File: client/src/main/java/io/split/engine/common/SplitAPI.java [hunk 7]
```diff
@@ -1,25 +1,27 @@
 package io.split.engine.common;
 
+import io.split.client.dtos.Split;
```
[the remark this answers is not in the document]
Comment (chillaq) [reply client/src/main/java/io/split/engine/common/SplitAPI.java:3]: will do

File: client/src/main/java/io/split/client/events/EventsSender.java (hunk 5 above)
Comment (chillaq) [reply client/src/main/java/io/split/client/events/EventsSender.java:22]: When I remove this one I get error for line 33:
_telemetryRuntimeProducer = checkNotNull(telemetryRuntimeProducer);

File: client/src/main/java/io/split/service/SplitHttpClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            Utils.forceClose(response);
+        }
+    }
+
+    public SplitHttpResponse post
+            (URI uri,
+             HttpEntity entity,
+             Map<String, String> additionalHeaders) throws IOException {
+        CloseableHttpResponse response = null;
+        long initTime = System.currentTimeMillis();
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/service/SplitHttpClientImpl.java:82]:
```suggestion
```

File: client/src/main/java/io/split/service/SplitHttpClientImpl.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                }
+            }
+            request.setEntity(entity);
+            request = (HttpPost) _requestDecorator.decorateHeaders(request);
+
+            response = _client.execute(request);
+
+            int status = response.getCode();
+
+            String statusMessage = new String("");
```
Comment (nmayorsplit) [inline client/src/main/java/io/split/service/SplitHttpClientImpl.java:97]:
```suggestion
            String statusMessage;
```

File: client/src/main/java/io/split/service/SplitHttpClientImpl.java (hunk 8 above)
Comment (chillaq) [reply client/src/main/java/io/split/service/SplitHttpClientImpl.java:82]: will do

File: client/src/main/java/io/split/service/SplitHttpClientImpl.java (hunk 9 above)
Comment (chillaq) [reply client/src/main/java/io/split/service/SplitHttpClientImpl.java:97]: If I remove this, I get error on line 104: statusMessage might not be initialized
httpResponse.statusMessage = statusMessage;

**Review by nmayorsplit — CHANGES_REQUESTED** [review]
I left some comments.

File: client/src/main/java/io/split/client/events/EventsSender.java (hunk 5 above)
Comment (nmayorsplit) [reply client/src/main/java/io/split/client/events/EventsSender.java:22]: ok, it'll be a change for the future to remove from the constructor parameters.

File: client/src/main/java/io/split/client/SplitClientConfig.java (hunk 3 above)
Comment (nmayorsplit) [reply client/src/main/java/io/split/client/SplitClientConfig.java:13]: yes, sure. But you should import each java.util, not use '*'.

Comment (chillaq) [reply client/src/main/java/io/split/client/SplitClientConfig.java:13]: Yep.

File: client/src/main/java/io/split/service/SplitHttpClientImpl.java (hunk 9 above)
Comment (chillaq) [reply client/src/main/java/io/split/service/SplitHttpClientImpl.java:96]: will do

**Review by nmayorsplit — APPROVED** [review]
LGTM

**Pull request merged**